### PR TITLE
[WIP] compress/decompress.go::Match modified

### PR DESCRIPTION
**Pull request by racytech**

### CODE REVIEW

**Review by AskAlexSharov — APPROVED** [review]
looks good

FYI: maybe can use golang's native fuzzing somehow. 
See FuzzCompress as an example.
can run it as: 
`go test -v -fuzz=FuzzCompress -fuzztime=10s ./compress`
it gives more meaningful randomness (by looking at coverage) and allows save in git repo some corner-cases corpus (example:  `txpool/testdata/fuzz/FuzzOnNewBlocks/` ) - which will run with usual tests (go test).

**Comment by AskAlexSharov** [comment]
@racytech just merged method `DisableFsync` for tests

File: compress/decompress.go [hunk 1]
```diff
@@ -680,20 +680,33 @@ func (g *Getter) Match(buf []byte) (bool, uint64) {
 			g.dataP++
 			g.dataBit = 0
 		}
-		if lenBuf != 0 {
+		if lenBuf != 0 || lenBuf != int(wordLen) {
```
Comment (AskAlexSharov) [inline compress/decompress.go:683]: if you found some bugs, please fix them in MatchPrefix method also.

**Comment by racytech** [comment]
When you say requestKey which is after latest key in file does it mean that I have to open reader and a writer at the same time or what does it mean? and for suffixes do I have to read entire word from file and compare it to given suffix?

**Comment by AskAlexSharov** [comment]
I mean: 
- file with keys {a, b, c}, search for key “f”. This test-case usually trigger edge-cases, like “forgotten -1” or “out of bound array read”, etc…


About suffixes: i didn’t mean any specific functionality (no new features). just “negative test-case”: search for “non existing key” and “search for existing suffix” may trigger different bugs - especially in method which does “partial key comparison”.

File: compress/decompress.go [hunk 2]
```diff
@@ -682,20 +692,33 @@ func (g *Getter) Match(buf []byte) (bool, uint64) {
...
 		}
-		return lenBuf == int(wordLen), g.dataP
+		if lenBuf == int(wordLen) {
+			return 0
+		}
+		if lenBuf < int(wordLen) {
+			return -1
+		}
+		if lenBuf > int(wordLen) {
+			return 1
```
Comment (AskAlexSharov) [inline compress/decompress.go:705]: seems forgot to restore internal state of cursor here
something like `g.dataP, g.dataBit = savePos, 0`

in my head - pattern of Match usage looks like: 
```
for getter.HasNext() {
   if getter.Match(prefix) {
       key, _ := getter.Next(nil)
       value, _ := getter.Next(nil)
       // process key/value ....
 }
}
```

**Review by AskAlexSharov — CHANGES_REQUESTED** [review]

File: compress/decompress.go [hunk 3]
```diff
@@ -712,28 +735,43 @@ func (g *Getter) Match(buf []byte) (bool, uint64) {
...
+	if lenBuf < int(wordLen) {
+		g.dataP, g.dataBit = savePos, 0
+		return -1
+	}
+	if lenBuf > int(wordLen) {
 		g.dataP, g.dataBit = savePos, 0
-		return false, savePos
+		return 1
 	}
 	g.dataP, g.dataBit = postLoopPos, 0
```
Comment (AskAlexSharov) [inline compress/decompress.go:773]: same here. seems it mutates getter position

File: compress/decompress.go (hunk 2 above)
Comment (racytech) [reply compress/decompress.go:705]: So the Match shouldn't move cursor to the next position if buf matches the word in file?

Comment (AskAlexSharov) [reply compress/decompress.go:705]: I don't know. Maybe you right (naming is always hard).
Have 2 different methods is fine for me. 

Then I need MatchPrefix to return `-1,0,1`. But there is 2 corner-cases: match of prefix (`ab` on `abcd`) and full-match of key (`abcd` on `abcd`). Maybe `MatchPrefix` can return `(cmp int, fullMatchOfKey bool, nextOffset uint64)`? We doing binary-search by this method in Erigon3.

**Comment by AskAlexSharov** [comment]
FYI: i created similar feature request - https://github.com/ledgerwatch/erigon-lib/issues/1053

File: compress/decompress.go [hunk 4]
```diff
@@ -712,32 +735,47 @@ func (g *Getter) Match(buf []byte) (bool, uint64) {
...
-	return true, postLoopPos
+	return 0
 }
 
 // MatchPrefix only checks if the word at the current offset has a buf prefix. Does not move offset to the next word.
-func (g *Getter) MatchPrefix(prefix []byte) bool {
+func (g *Getter) MatchPrefix(prefix []byte, nf uint64) (cmp int, fullMatchOfKey bool, nextOffset uint64) {
```
Comment (AskAlexSharov) [inline compress/decompress.go:778]: What is “nf”?

**Review by AskAlexSharov — APPROVED** [review]

File: compress/decompress.go (hunk 4 above)
Comment (racytech) [reply compress/decompress.go:778]: This will be removed. I was testing that the next offset is correct, so I had to pass returned value again to function.

**Review by AskAlexSharov — APPROVED** [review]
Let’s return wordLenn from SkipUncompressed tool

Let’s add method MatchPrefixUncompressed?

**Comment by racytech** [comment]
Will create a new PR, since this one isn't right
